SF-2942 Display warning message when editor does not support translation language

kylebuss opened the pull request:
Added `app-notice` to `editor.component.html` to notify users to use Firefox if translation language is Korean, Japanese, or Chinese and the browser they are using is not Firefox.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/web-xforge/2699)
<!-- Reviewable:end -->

RaymondLuong3 reviewed (changes requested):
Thanks for the update. I've left some comments.

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-:-O5tgETzAni8h60cnU6n:b-onarjt)* status: all files reviewed, 5 unresolved discussions (waiting on @kylebuss)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 531 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5tcqFU5gqCqtIl7Lyt:-O5tcqFU5gqCqtIl7Lyu:bvbsk1f) ([raw file](https://github.com/sillsdev/web-xforge/blob/d0873227cd05aca23e1934f27449af80350943bb/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L531)):*
> ```XML
> 
>   get browserWarningMessage(): string {
>     return translate('editor.browser_warning_banner', { firefoxLink: getLinkHTML('Firefox', 'https://firefox.com') });
> ```

You can use the firefox link from `xforge-common/utils.ts`



_Code quote:_
```XML
    return translate('editor.browser_warning_banner', { firefoxLink: getLinkHTML('Firefox', 'https://firefox.com') });
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 613 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5tg3kR21I9NQ1ZjafU:-O5tg3kR21I9NQ1ZjafV:bpkz2r8) ([raw file](https://github.com/sillsdev/web-xforge/blob/d0873227cd05aca23e1934f27449af80350943bb/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L613)):*
> ```XML
>   }
> 
>   get showBrowserWarningBanner(): boolean {
> ```

How about changing this to writingSystemWarningBanner?



_Code quote:_
```XML
  get showBrowserWarningBanner(): boolean {
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 617 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5tdxmyAYr4LopBfIsU:-O5tdxmz4DJoKR-4N8pt:b5r7o8w) ([raw file](https://github.com/sillsdev/web-xforge/blob/d0873227cd05aca23e1934f27449af80350943bb/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L617)):*
> ```XML
>     return (
>       writingSystemTag != null &&
>       this.translationWarningLanguages.includes(writingSystemTag) &&
> ```

We should be more specific when we parse writing system tags. There are so many writing system tags, there is bound to be one that looks like `bko', but we don't want to mistake that to be Korean.
In this case, you might have to write a Regex for it. Off the top of my head, something like `/^(ko|cmn|ja)_?/`. Start with 'ko' or 'cmn' or 'ja' with 0 or more underscores.



_Code quote:_
```XML
      writingSystemTag != null &&
      this.translationWarningLanguages.includes(writingSystemTag) &&
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts` line 4474 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5tfuqM98aNuX3dP0Yo:-O5tfuqM98aNuX3dP0Yp:bbl7qmu) ([raw file](https://github.com/sillsdev/web-xforge/blob/d0873227cd05aca23e1934f27449af80350943bb/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts#L4474)):*
> ```XML
>   }
> 
>   get browserWarningBanner(): DebugElement {
> ```

What about changing this to writingSystemWarningBanner?



_Code quote:_
```XML
  get browserWarningBanner(): DebugElement {
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 225 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5tdBW9EZpzNB7YZCp0:-O5tdBW9EZpzNB7YZCp1:b-5z83gi) ([raw file](https://github.com/sillsdev/web-xforge/blob/d0873227cd05aca23e1934f27449af80350943bb/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L225)):*
> ```JSON
>     "add_comment": "Add Comment",
>     "anonymous": "Anonymous",
>     "browser_warning_banner": "Translating to this language in Chrome could result in errors displaying in the editor. We recommend using {{ firefoxLink }}.",
> ```

This message can be improved. How about "Your browser does not support editing in this language. We recommend using firefox."



_Code quote:_
```JSON
"Translating to this language in Chrome could result in errors displaying in the editor. We recommend using {{ firefoxLink }}.",
```


<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (changes requested):
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 3 unresolved discussions (waiting on @kylebuss)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 100 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5y4ZRK6G4JN6E2XF54:-O5y4ZRK6G4JN6E2XF55:b-iivo7a) ([raw file](https://github.com/sillsdev/web-xforge/blob/2d9d367e320dde59d679af1596a10868bb48f480/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L100)):*
> ```XML
> import { UserService } from 'xforge-common/user.service';
> import { filterNullish } from 'xforge-common/util/rxjs-util';
> import { getLinkHTML, issuesEmailTemplate, objectId, browserLinks, isGecko } from 'xforge-common/utils';
> ```

Nit: these imports are out of order. My environment will order them automatically. I can't tell which extension is doing that for me.



_Code quote:_
```XML
import { getLinkHTML, issuesEmailTemplate, objectId, browserLinks, isGecko 
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 254 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5y5Dr6AAbCEdDeC6ye:-O5y5Dr6AAbCEdDeC6yf:b-w4rcle) ([raw file](https://github.com/sillsdev/web-xforge/blob/2d9d367e320dde59d679af1596a10868bb48f480/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L254)):*
> ```XML
>   private readonly fabDiameter = 40;
>   readonly fabVerticalCushion = 5;
>   readonly writingSystemWarningRegex = /^(ko|cmn|ja)_?$/;
> ```

You can put this at the top of this file as a const. You'll see other examples of that in this file.
This will also need updated. I did not account for strings that look like `jad_mn`. Try this



_Code quote (i):_
```XML
  readonly writingSystemWarningRegex = /^(ko|cmn|ja)_?$/;
```



_Code snippet (ii):_
```XML
/^(ja|ko|cmn)(?![a-z])/
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 225 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5tdBW9EZpzNB7YZCp0:-O5y6YCX5QRZeWWoSFB5:b-nkjwxu) ([raw file](https://github.com/sillsdev/web-xforge/blob/d0873227cd05aca23e1934f27449af80350943bb/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L225)):*
<details><summary><i>Previously, kylebuss (Kyle Buss) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Sorry, you should remove the 'firefox' part and just have the link there which already will appear as 'firefox'


<!-- Sent from Reviewable.io -->

kylebuss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 3 unresolved discussions (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 100 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5y4ZRK6G4JN6E2XF54:-O5yNBYLDIljifCeVWBD:b878mhk) ([raw file](https://github.com/sillsdev/web-xforge/blob/2d9d367e320dde59d679af1596a10868bb48f480/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L100)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Nit: these imports are out of order. My environment will order them automatically. I can't tell which extension is doing that for me.
</blockquote></details>

Surprised this isn't caught by `ng lint`. If you figure out what extension is sorting imports let me know so I can add it.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 254 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5y5Dr6AAbCEdDeC6ye:-O5yV5OT7WiwhFhH8YtW:b8ht17d) ([raw file](https://github.com/sillsdev/web-xforge/blob/2d9d367e320dde59d679af1596a10868bb48f480/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L254)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

You can put this at the top of this file as a const. You'll see other examples of that in this file.
This will also need updated. I did not account for strings that look like `jad_mn`. Try this
</blockquote></details>

Are we trying to include other writing system tags that would be affected by this or limiting it to `ja|ko|cmn`?
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 225 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5tdBW9EZpzNB7YZCp0:-O5yWQXo-6084rquUvTI:b-h1158j) ([raw file](https://github.com/sillsdev/web-xforge/blob/d0873227cd05aca23e1934f27449af80350943bb/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L225)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Sorry, you should remove the 'firefox' part and just have the link there which already will appear as 'firefox'
</blockquote></details>

My bad, I thought I had done that.


<!-- Sent from Reviewable.io -->

kylebuss commented:
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 254 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5y5Dr6AAbCEdDeC6ye:-O612oldAsoupfRc9Tf6:b-x7ukpp) ([raw file](https://github.com/sillsdev/web-xforge/blob/2d9d367e320dde59d679af1596a10868bb48f480/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L254)):*
<details><summary><i>Previously, kylebuss (Kyle Buss) wrote…</i></summary><blockquote>

Are we trying to include other writing system tags that would be affected by this or limiting it to `ja|ko|cmn`?
</blockquote></details>

`/^(ko|ja|cmn)$/` will only return true if it exactly matches one of the options, however if we are trying to match additional tags that start with `ko|ja|cmn` and have additional characters the regex will need updated. 


<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (changes requested):
One final comment and then this should be good to go.

> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-:-O61zTXlDQFQuxyV_4ry:b6i1ibm)* status: all files reviewed, 1 unresolved discussion (waiting on @kylebuss)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 254 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5y5Dr6AAbCEdDeC6ye:-O61yWxY8dELuSCuS48V:b-mxj3mc) ([raw file](https://github.com/sillsdev/web-xforge/blob/2d9d367e320dde59d679af1596a10868bb48f480/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L254)):*
<details><summary><i>Previously, kylebuss (Kyle Buss) wrote…</i></summary><blockquote>

`/^(ko|ja|cmn)$/` will only return true if it exactly matches one of the options, however if we are trying to match additional tags that start with `ko|ja|cmn` and have additional characters the regex will need updated.
</blockquote></details>

Writing system tags will typically look like `xx`, however, there are optional forms that include the locale and the script (xx_locale_script). In this case, it is not likely that there will be a locale or script attached. But to be safe, lets make sure that something like `ja_nor` will be correctly identified as a language we cannot support.


<!-- Sent from Reviewable.io -->

kylebuss commented:
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 254 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5y5Dr6AAbCEdDeC6ye:-O61ztQiDfEWhiI9pXFe:bg4k9a5) ([raw file](https://github.com/sillsdev/web-xforge/blob/2d9d367e320dde59d679af1596a10868bb48f480/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L254)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Writing system tags will typically look like `xx`, however, there are optional forms that include the locale and the script (xx_locale_script). In this case, it is not likely that there will be a locale or script attached. But to be safe, lets make sure that something like `ja_nor` will be correctly identified as a language we cannot support.
</blockquote></details>

Okay. I tested this `/^(ko|ja|cmn)(_{1}[a-z]*)*$/` on [regex](https://regex101.com/) and believe this is what we are looking for. Let me know if you think we need something different.


<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 1 unresolved discussion (waiting on @kylebuss)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 254 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5y5Dr6AAbCEdDeC6ye:-O625wiw590jBPhpmHyc:b-rl1j4t) ([raw file](https://github.com/sillsdev/web-xforge/blob/2d9d367e320dde59d679af1596a10868bb48f480/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L254)):*
<details><summary><i>Previously, kylebuss (Kyle Buss) wrote…</i></summary><blockquote>

Okay. I tested this `/^(ko|ja|cmn)(_{1}[a-z]*)*$/` on [regex](https://regex101.com/) and believe this is what we are looking for. Let me know if you think we need something different.
</blockquote></details>

Hmm, we don't need to be so specific, and it would trip on a string like `ko_sth_chn` if such a ws tag exists. What I suggested above should work. `/^(ja|ko|cmn)(?![a-z])/` will that any string that looks like `ko` or `ko_xx` will be matched, but not `kop`


<!-- Sent from Reviewable.io -->

kylebuss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 254 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O5y5Dr6AAbCEdDeC6ye:-O62kHOl0olQ6MGIsa1y:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/2d9d367e320dde59d679af1596a10868bb48f480/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L254)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Hmm, we don't need to be so specific, and it would trip on a string like `ko_sth_chn` if such a ws tag exists. What I suggested above should work. `/^(ja|ko|cmn)(?![a-z])/` will that any string that looks like `ko` or `ko_xx` will be matched, but not `kop`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (approved):
I can confirm there is such a ws as `cmn-Hans-CN` so I'm glad we took the time to get the regex for the ws just right. Please update the Jira issue with acceptance tests.

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-:-O62zC7O-BlM0Dv6V0TK:b-ptkqp)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @kylebuss)



<!-- Sent from Reviewable.io -->

Nateowami reviewed (changes requested):
Holding

Nateowami commented:
The task says "Warn translators/admins". Really it should warn the user only when the user can edit. I can't see anything in this PR that limit the message to times when the user can edit the text.

kylebuss commented:
Good catch! I'll add a check that user has edit permissions before displaying the message.





<!-- Sent from Reviewable.io -->

Nateowami commented:
@kylebuss Consider the `canEdit` getter

Nateowami commented:
I have dismissed my review. @RaymondLuong3 can review the changes.

RaymondLuong3 reviewed (changes requested):
Looks good. Just one minor comment.

> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-:-O6b7LWu7KuaE17WNpfg:b-izhvug)* status: all files reviewed, 1 unresolved discussion (waiting on @kylebuss and @Nateowami)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts` line 272 at r5](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O6b7ENo93XQUYTv0tEg:-O6b7ENo93XQUYTv0tEh:baqu4h6) ([raw file](https://github.com/sillsdev/web-xforge/blob/28c6b5dc664f9fe89dafb58b7ececcabc0f403f2/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts#L272)):*
> ```XML
>       writingSystem: { tag: 'ko' },
>       translateConfig: defaultTranslateConfig
>     });
> ```

Could you add a comment here explaining why this text is not editable?


<!-- Sent from Reviewable.io -->

Nateowami reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 2 unresolved discussions (waiting on @kylebuss)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 164 at r5](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O6bGi6Z-nRL-LTZmmnk:-O6bGi6Z-nRL-LTZmmnl:b-cs1nfm) ([raw file](https://github.com/sillsdev/web-xforge/blob/28c6b5dc664f9fe89dafb58b7ececcabc0f403f2/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L164)):*
> ```XML
> 
> const PUNCT_SPACE_REGEX = /^(?:\p{P}|\p{S}|\p{Cc}|\p{Z})+$/u;
> const WRITING_SYSTEM_WARNING_REGEX = /^(ja|ko|cmn)(?![a-z])/;
> ```

Can you explain why this is right? Maybe you've done the homework and figured it out, but I can't follow. It seems to be missing things like `zh`, `zho`, `chi`. Do you know for certain what format of ISO code is being used, and therefore an exhaustive list of language codes that refer to CJK languages?

I would probably suggest something like:
``` ts
function isCJKCode(code: string): boolean {
  const leadingPartOfCode = code.split('-')[0]; // or should we be splitting on : or something else?
  return ['ja', 'ko', 'cmn', 'zh'].includes(leadingPartOfCode);
}
```
But that's probably not sufficient.

I believe we get the language code right here: https://github.com/sillsdev/web-xforge/blob/19b30a30c542b517d89915b8fa1d1d04e82e3306/src/SIL.XForge.Scripture/Services/ParatextService.cs#L2213

Though you'd have to figure out exactly what format we are getting from PT to figure out the format of codes we should check for.


<!-- Sent from Reviewable.io -->

kylebuss commented:
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 164 at r5](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O6bGi6Z-nRL-LTZmmnk:-O6bkin9BQ-ejqZWDDBz:bikjdm9) ([raw file](https://github.com/sillsdev/web-xforge/blob/28c6b5dc664f9fe89dafb58b7ececcabc0f403f2/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L164)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

Can you explain why this is right? Maybe you've done the homework and figured it out, but I can't follow. It seems to be missing things like `zh`, `zho`, `chi`. Do you know for certain what format of ISO code is being used, and therefore an exhaustive list of language codes that refer to CJK languages?

I would probably suggest something like:
``` ts
function isCJKCode(code: string): boolean {
  const leadingPartOfCode = code.split('-')[0]; // or should we be splitting on : or something else?
  return ['ja', 'ko', 'cmn', 'zh'].includes(leadingPartOfCode);
}
```
But that's probably not sufficient.

I believe we get the language code right here: https://github.com/sillsdev/web-xforge/blob/19b30a30c542b517d89915b8fa1d1d04e82e3306/src/SIL.XForge.Scripture/Services/ParatextService.cs#L2213

Though you'd have to figure out exactly what format we are getting from PT to figure out the format of codes we should check for.
</blockquote></details>

From what I can tell we use ISO 639-1 for Korean and Japanese but ISO 639-3 for [Chinese languages](https://iso639-3.sil.org/code/zho). I've grabbed all the Chinese ISO 639-3 codes and updated your code suggestions return statement below. Do you think we should include the sign language codes or is it safe to exclude them?

I figured out you can view the language code that will be used in the Paratext Registry Project Edit screen by clicking Advanced and viewing the Identifier:

![Identifier.png](https://files.reviewableusercontent.io/images/reviewable/174831353/feYJL2cMPZZTfp2al5RlcmhN/Identifier.png)

The regex was doing basically the same thing as your example. It was matching if the language code was the beginning of the string. Your code suggestion is probably easier to read/understand.

Updating your return to include a more exhaustive list of codes:
`return ['ko', 'kor', 'ja', 'jpn', 'cmn', 'czh', 'cdo', 'cjy', 'cmn', 'cpx', 'czh', 'czo', 'gan', 'hak', 'hsn', 'lzh', 'mnp', 'nan', 'quu', 'yue', 'cnp', 'csp', 'cpi', 'lzh', 'lpz', 'wuu', 'zh'].includes(leadingPartOfCode);`

Let me know if you think of anything else.


<!-- Sent from Reviewable.io -->

kylebuss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 2 unresolved discussions (waiting on @Nateowami and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts` line 272 at r5](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O6b7ENo93XQUYTv0tEg:-O6c-dUZ815gjT91E92D:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/28c6b5dc664f9fe89dafb58b7ececcabc0f403f2/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts#L272)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Could you add a comment here explaining why this text is not editable?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (changes requested):
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 3 unresolved discussions (waiting on @kylebuss and @Nateowami)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 641 at r6](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7BGvro72TtdzciKQfU:-O7BGvro72TtdzciKQfV:b-esit64) ([raw file](https://github.com/sillsdev/web-xforge/blob/4d62012636e830f7834d4251bc2ca3ccef4c3f9e/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L641)):*
> ```XML
>   get writingSystemWarningBanner(): boolean {
>     const writingSystemTag = this.projectDoc?.data?.writingSystem.tag;
>     const languageCode = writingSystemTag?.split('-')[0] ?? '';
> ```

Actually, language tags are in the format code_



_Code quote:_
```XML
    const languageCode = writingSystemTag?.split('-')[0] ?? '';
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts` line 288 at r6](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7BGnsz2Ko2k7MeFPYT:-O7BGnsz2Ko2k7MeFPYU:b-tphjuv) ([raw file](https://github.com/sillsdev/web-xforge/blob/4d62012636e830f7834d4251bc2ca3ccef4c3f9e/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts#L288)):*
> ```XML
>     expect(env.component.canEdit).toBe(false);
>     expect(env.component.showNoEditPermissionMessage).toBe(true); //false
>     expect(env.noChapterEditPermissionMessage).not.toBeNull(); // null
> ```

I think you left some comments in here by accident.



_Code quote:_
```XML
    expect(env.component.hasChapterEditPermission).toBe(false); // undefined
    expect(env.component.canEdit).toBe(false);
    expect(env.component.showNoEditPermissionMessage).toBe(true); //false
    expect(env.noChapterEditPermissionMessage).not.toBeNull(); // null
```


<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (commented):
@Nateowami It looks like you still have requested changes on this PR.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-:-O7BHh2C02mLzkHmZ50j:bric05r)* status: all files reviewed, 2 unresolved discussions (waiting on @kylebuss and @Nateowami)



<!-- Sent from Reviewable.io -->

kylebuss commented:
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 641 at r6](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7BGvro72TtdzciKQfU:-O7BJQ1uEVdSfeL4_ahz:bffb6a5) ([raw file](https://github.com/sillsdev/web-xforge/blob/4d62012636e830f7834d4251bc2ca3ccef4c3f9e/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L641)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Actually, language tags are in the format code_
</blockquote></details>

I'm seeing the `-` being used in the Paratext Registry Identifier which is where we get the code from.

Thoughts on`writingSystemTag?.split('-').split('_')[0] ?? ''` to ensure both `-` and `_` are handled?


<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 2 unresolved discussions (waiting on @kylebuss and @Nateowami)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 641 at r6](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7BGvro72TtdzciKQfU:-O7BKVQz01C0R6iDqOlN:b-e9f6ya) ([raw file](https://github.com/sillsdev/web-xforge/blob/4d62012636e830f7834d4251bc2ca3ccef4c3f9e/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L641)):*
<details><summary><i>Previously, kylebuss (Kyle Buss) wrote…</i></summary><blockquote>

I'm seeing the `-` being used in the Paratext Registry Identifier which is where we get the code from.

Thoughts on`writingSystemTag?.split('-').split('_')[0] ?? ''` to ensure both `-` and `_` are handled?
</blockquote></details>

I actually retracted the comment. I think you have it right.


<!-- Sent from Reviewable.io -->

Nateowami reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 2 unresolved discussions (waiting on @kylebuss)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 641 at r6](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7BGvro72TtdzciKQfU:-O7FUNtkAzAAFv1JZoJR:b-e6su44) ([raw file](https://github.com/sillsdev/web-xforge/blob/4d62012636e830f7834d4251bc2ca3ccef4c3f9e/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L641)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

I actually retracted the comment. I think you have it right.
</blockquote></details>

Using `.split()` twice would not work, because the second call would be on an array. You would need to do `string.split(/[-_]/)` if you wanted to do this.


<!-- Sent from Reviewable.io -->

kylebuss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 2 unresolved discussions (waiting on @Nateowami and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts` line 288 at r6](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7BGnsz2Ko2k7MeFPYT:-O7FyUrJAhuWS9MY-j3p:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/4d62012636e830f7834d4251bc2ca3ccef4c3f9e/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts#L288)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

I think you left some comments in here by accident.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (approved):
> Reviewed 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 1 unresolved discussion (waiting on @Nateowami)



<!-- Sent from Reviewable.io -->

Nateowami reviewed (changes requested):
> Reviewed 2 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 3 unresolved discussions (waiting on @kylebuss)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.html` line 118 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7olsu_D2yA9scblt0f:-O7olsu_D2yA9scblt0g:bwj0o0w) ([raw file](https://github.com/sillsdev/web-xforge/blob/f0d3f1bb65488ab9a558af9f2d1a92a151fb33be/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.html#L118)):*
> ```HTML
>                   @if (writingSystemWarningBanner) {
>                     <app-notice icon="warning" type="warning" class="writing-system-warning-banner">
>                       <div>
> ```

Why is there a div wrapping this?
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 164 at r5](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O6bGi6Z-nRL-LTZmmnk:-O7oofPg-8ej_YgOof77:b-njeskw) ([raw file](https://github.com/sillsdev/web-xforge/blob/28c6b5dc664f9fe89dafb58b7ececcabc0f403f2/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L164)):*
<details><summary><i>Previously, kylebuss (Kyle Buss) wrote…</i></summary><blockquote>

From what I can tell we use ISO 639-1 for Korean and Japanese but ISO 639-3 for [Chinese languages](https://iso639-3.sil.org/code/zho). I've grabbed all the Chinese ISO 639-3 codes and updated your code suggestions return statement below. Do you think we should include the sign language codes or is it safe to exclude them?

I figured out you can view the language code that will be used in the Paratext Registry Project Edit screen by clicking Advanced and viewing the Identifier:

![Identifier.png](https://files.reviewableusercontent.io/images/reviewable/174831353/feYJL2cMPZZTfp2al5RlcmhN/Identifier.png)

The regex was doing basically the same thing as your example. It was matching if the language code was the beginning of the string. Your code suggestion is probably easier to read/understand.

Updating your return to include a more exhaustive list of codes:
`return ['ko', 'kor', 'ja', 'jpn', 'cmn', 'czh', 'cdo', 'cjy', 'cmn', 'cpx', 'czh', 'czo', 'gan', 'hak', 'hsn', 'lzh', 'mnp', 'nan', 'quu', 'yue', 'cnp', 'csp', 'cpi', 'lzh', 'lpz', 'wuu', 'zh'].includes(leadingPartOfCode);`

Let me know if you think of anything else.
</blockquote></details>

I am not convinced that this is the right list of language codes, but it's acceptable at least for now until we have a better way of detecting the script.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 625 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7oo6Du76qMGgZDPDl_:-O7oo6Dv3O912zwPlD5x:b-a6pktl) ([raw file](https://github.com/sillsdev/web-xforge/blob/f0d3f1bb65488ab9a558af9f2d1a92a151fb33be/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L625)):*
> ```XML
>     const languageCode = writingSystemTag?.split(/-_/)[0] ?? '';
>     const unsupportedLanguageCode = UNSUPPORTED_LANGUAGE_CODES.includes(languageCode);
>     return !isGecko() && writingSystemTag != null && unsupportedLanguageCode && this.canEdit;
> ```

The issue says it doesn't work in Chrome, not that it only works in Firefox.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 251 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7olN6h6-6I3_fLTaWi:-O7olN6h6-6I3_fLTaWj:b-x2162t) ([raw file](https://github.com/sillsdev/web-xforge/blob/f0d3f1bb65488ab9a558af9f2d1a92a151fb33be/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L251)):*
> ```JSON
>     "add_comment": "Add Comment",
>     "anonymous": "Anonymous",
>     "browser_warning_banner": "Your browser does not support editing in this language. We recommend using {{ firefoxLink }}.",
> ```

1. This message is not accurate. It's not the browser that doesn't support it, it's our editor.
2. Is Firefox actually the only alternative?


<!-- Sent from Reviewable.io -->

kylebuss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 3 unresolved discussions (waiting on @Nateowami)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.html` line 118 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7olsu_D2yA9scblt0f:-O81xwgGAJOn__226PlO:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/f0d3f1bb65488ab9a558af9f2d1a92a151fb33be/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.html#L118)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

Why is there a div wrapping this?
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 625 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7oo6Du76qMGgZDPDl_:-O81xmYPAnx-g_G9vL6q:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/f0d3f1bb65488ab9a558af9f2d1a92a151fb33be/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L625)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

The issue says it doesn't work in Chrome, not that it only works in Firefox.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 251 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O7olN6h6-6I3_fLTaWi:-O820sXzEyiCBNhBWk1G:b-l1c1em) ([raw file](https://github.com/sillsdev/web-xforge/blob/f0d3f1bb65488ab9a558af9f2d1a92a151fb33be/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L251)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

1. This message is not accurate. It's not the browser that doesn't support it, it's our editor.
2. Is Firefox actually the only alternative?
</blockquote></details>

I am not finding anything that suggests the issue exists in Safari. I will include it with the message update.


<!-- Sent from Reviewable.io -->

Nateowami reviewed (changes requested):
> Reviewed 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 2 unresolved discussions (waiting on @kylebuss)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 247 at r8](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O82USLOBwJEZd3JmE3Z:-O82USLOBwJEZd3JmE3_:b-k52grk) ([raw file](https://github.com/sillsdev/web-xforge/blob/1f58a70f203ed2a35ec5d18b7ce3107a542b4977/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L247)):*
> ```JSON
>     "add_comment": "Add Comment",
>     "anonymous": "Anonymous",
>     "browser_warning_banner": "The editor is not compatible with the current browser for the projects language. We recommend translating this project in {{ firefoxLink }} or {{ safari }}.",
> ```

Should this be "project language"? I could see how maybe it could be "project's language", but not "projects language"
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/utils.ts` line 73 at r8](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O82VTH58a0WY03e71cZ:-O82VTH58a0WY03e71c_:btogrwj) ([raw file](https://github.com/sillsdev/web-xforge/blob/1f58a70f203ed2a35ec5d18b7ce3107a542b4977/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/utils.ts#L73)):*
> ```XML
> }
> 
> export function isWebkit(): boolean {
> ```

We don't need this method because it's not used. I'm guessing you intended to add it for completeness, but a) I don't think that's necessary, and b) I'm not sure it really is complete.


<!-- Sent from Reviewable.io -->

kylebuss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: all files reviewed, 2 unresolved discussions (waiting on @Nateowami)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 247 at r8](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O82USLOBwJEZd3JmE3Z:-O82akr_7ezFDOZ_Wxzi:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/1f58a70f203ed2a35ec5d18b7ce3107a542b4977/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L247)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

Should this be "project language"? I could see how maybe it could be "project's language", but not "projects language"
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/utils.ts` line 73 at r8](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O82VTH58a0WY03e71cZ:-O82alKm0ayaLBOPst-a:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/1f58a70f203ed2a35ec5d18b7ce3107a542b4977/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/utils.ts#L73)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

We don't need this method because it's not used. I'm guessing you intended to add it for completeness, but a) I don't think that's necessary, and b) I'm not sure it really is complete.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kylebuss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: 3 of 5 files reviewed, 1 unresolved discussion (waiting on @Nateowami)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 247 at r8](https://reviewable.io/reviews/sillsdev/web-xforge/2699#-O82USLOBwJEZd3JmE3Z:-O82pIpO4u0IUVA7-Tpj:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/1f58a70f203ed2a35ec5d18b7ce3107a542b4977/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L247)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

Can you change "translating" to "editing"?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Nateowami reviewed (approved):
> Reviewed 1 of 2 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2699)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @kylebuss)



<!-- Sent from Reviewable.io -->
